app/services/record_declarations/base.rb

```diff
@@ -106,9 +97,63 @@ def create_record!
       end
     end
 
+    def valid_courses_for_user
+      valid_courses = []
+      valid_courses << "ecf-mentor" if user.mentor?
+      valid_courses << "ecf-induction" if user.early_career_teacher?
+      valid_courses += NPQCourse.all.map(&:identifier) if user.npq?
+      valid_courses
+    end
+
+    def profile_exists
+      return if errors.any?
+
+      unless user_profile
+        errors.add(:user_profile, I18n.t(:invalid_participant))
+      end
+    end
+
+    def date_has_the_right_format
+      return if declaration_date.blank?
+
+      errors.add(:declaration_date, I18n.t(:invalid_declaration_date)) unless declaration_date.match(RFC3339_DATE_REGEX)
+      parsed_date
+    rescue StandardError
+      errors.add(:declaration_date, I18n.t(:invalid_declaration_date))
+    end
+
     def validate_provider!
       # TODO: Remove the nil? check and fix the test setup so that they build the school cohort, partnership and give us back the actual lead_provider.
       raise ActionController::ParameterMissing, I18n.t(:invalid_participant) unless actual_lead_provider.nil? || lead_provider_from_token == actual_lead_provider
     end
+
+    def validate_milestone!
+      unless milestone.start_date < parsed_date
+        raise ActionController::ParameterMissing, I18n.t(:declaration_before_milestone_start)
+      end
+
+      unless milestone.milestone_date > parsed_date
+        raise ActionController::ParameterMissing, I18n.t(:declaration_after_milestone_cutoff)
+      end
+    end
+
+    def milestone
+      unless schedule
+        raise ActionController::ParameterMissing, I18n.t(:schedule_missing)
+      end
+
+      declaration_to_milestone_map[declaration_type.to_sym]
+    end
+
+    def declaration_to_milestone_map
+      {
+        "started": schedule.milestones[0],
+        "retained-1": schedule.milestones[1],
+        "retained-2": schedule.milestones[2],
+        "retained-3": schedule.milestones[3],
+        "retained-4": schedule.milestones[4],
+        "completed": schedule.milestones.last,
+      }
+    end
```

Comment:
Probably easier to follow is you do

`declaration_to_milestone_map[declaration_type.underscore.to_sym]`

and have

`def declaration_to_milestone_map
{
started: schedule.milestone[0],
retained_1: schedule.milestons[1],
.
.
.
etc.
}`